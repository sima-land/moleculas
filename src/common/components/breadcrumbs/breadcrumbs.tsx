import { type SVGAttributes, type ReactNode, useState, useRef, useEffect } from 'react';
import { type BreadcrumbItemProps, type BreadcrumbProps } from './types';
import { Link, type LinkProps } from '@sima-land/ui-nucleons/link';
import { Plate } from '@sima-land/ui-nucleons/plate';
import { defineSlots } from '@sima-land/ui-nucleons/helpers';
import { useOutsideClick } from '@sima-land/ui-nucleons/hooks';
import { CSSTransition } from 'react-transition-group';
import classNames from 'classnames';
import styles from './breadcrumbs.m.scss';

/**
 * Список "хлебных крошек".
 * @param props Свойства.
 * @return Элемент.
 */
export function Breadcrumbs({
  children,
  className,
  'data-testid': testId = 'breadcrumbs',
  ...restProps
}: BreadcrumbProps) {
  return (
    <div {...restProps} className={classNames(styles.root, className)} data-testid={testId}>
      {children}
    </div>
  );
}

/**
 * Элемент списка "хлебных крошек".
 * @param props Свойства.
 * @return Элемент.
 */
export function BreadcrumbsItem({
  active,
  children,
  'data-testid': testId = 'breadcrumb',
  ...restProps
}: BreadcrumbItemProps) {
  const { body, dropdown } = defineSlots(children, {
    body: BreadcrumbsItemBody,
    dropdown: BreadcrumbsItemDropdown,
  });

  const [open, setOpen] = useState(false);

  const ref = useRef<HTMLDivElement>(null);

  useOutsideClick(ref, () => setOpen(false));

  // ВАЖНО: выносим кусок верстки чтобы использовать его в двух местах
  const name = (
    <span className={styles.name}>
      <Link color='basic-gray87' {...(active && { pseudo: true })} {...restProps}>
        {body}
      </Link>
      {dropdown && <DownSVG className={styles.icon} onClick={() => setOpen(a => !a)} />}
    </span>
  );

  return (
    <div className={styles.item} data-testid={testId}>
      {/* ВАЖНО: inner нужен для того чтобы позиционировать dropdown (иначе надо учитывать разделитель "/") */}
      <div className={styles.inner}>
        {name}
        <CSSTransition
          in={dropdown && open}
          timeout={200}
          unmountOnExit
          classNames={{
            enter: styles['fade-enter'],
            enterActive: styles['fade-enter-active'],
            exit: styles['fade-exit'],
            exitActive: styles['fade-exit-active'],
          }}
        >
          <Plate ref={ref} rounds='m' shadow='z3' className={styles.dropdown}>
            {name}
            <div className={styles['dropdown-body']}>{dropdown}</div>
          </Plate>
        </CSSTransition>
      </div>
    </div>
  );
}

BreadcrumbsItem.Body = BreadcrumbsItemBody;
BreadcrumbsItem.Dropdown = BreadcrumbsItemDropdown;
BreadcrumbsItem.Sibling = BreadcrumbsItemSibling;

/**
 * Основное содержимое "хлебной крошки".
 * @param props Свойства.
 * @return Элемент.
 */
export function BreadcrumbsItemBody({ children }: { children?: ReactNode }) {
  return <>{children}</>;
}

/**
 * Содержимое выпадающего списка "хлебной крошки".
 * @param props Свойства.
 * @return Элемент.
 */
export function BreadcrumbsItemDropdown({
  loading,
  children,
  onOpen,
}: {
  children?: ReactNode;
  loading?: boolean;
  onOpen?: VoidFunction;
}) {
  useEffect(() => onOpen?.(), []);

  return (
    <>
      {loading ? (
        <div className={styles.stubs}>
          {/* ВАЖНО: по гайдам нужно именно 8 штук */}
          <div className={styles.stub} />
          <div className={styles.stub} />
          <div className={styles.stub} />
          <div className={styles.stub} />
          <div className={styles.stub} />
          <div className={styles.stub} />
          <div className={styles.stub} />
          <div className={styles.stub} />
        </div>
      ) : (
        children
      )}
    </>
  );
}

/**
 * Соседняя "хлебная крошка", выводится в выпадающем списке.
 * @param props Свойства.
 * @return Элемент.
 */
export function BreadcrumbsItemSibling({
  children,
  className,
  active,
  'data-testid': testId = 'breadcrumb:sibling',
  ...restProps
}: LinkProps & { active?: boolean }) {
  return (
    <Link
      color='basic-gray87'
      className={classNames(styles.sibling, className)}
      data-testid={testId}
      {...(active && { disabled: true, pseudo: true })}
      {...restProps}
    >
      {children}
    </Link>
  );
}

/** @inheritdoc */
function DownSVG(props: SVGAttributes<SVGSVGElement>) {
  return (
    <svg width='16' height='16' viewBox='0 0 16 16' fill='none' {...props}>
      <path d='M4 6L8 10L12 6' strokeLinecap='round' strokeLinejoin='round' />
    </svg>
  );
}
