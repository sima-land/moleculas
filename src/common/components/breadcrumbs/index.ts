export type { BreadcrumbProps, BreadcrumbItemProps } from './types';
export { Breadcrumbs, BreadcrumbsItem } from './breadcrumbs';
