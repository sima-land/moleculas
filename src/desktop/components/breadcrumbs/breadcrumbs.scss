@use 'node_modules/@dev-dep/ui-nucleons/colors';

$popup-padding: 16px;
$line-gutter: 12px;

.root {
  display: flex;
  flex-wrap: wrap;
  margin: 0;
  padding: 0;
  border: 0;
  outline: 0;
  font-size: 12px;
  line-height: 16px;
  user-select: none;
  margin-bottom: -$line-gutter;
}

.breadcrumb {
  display: flex;
  flex-wrap: nowrap;
  position: relative;
  white-space: nowrap;
  margin-bottom: $line-gutter;
  &:not(:last-child)::after {
    content: '/';
    display: block;
    margin: 0 16px;
  }
}

.toggle-icon {
  display: block;
  cursor: pointer;
  &.close {
    transform: scaleY(-1);
  }
}

.popup {
  min-width: 180px;
  padding: $popup-padding;
  position: absolute;
  top: -$popup-padding;
  left: -$popup-padding;
  z-index: 2;
}

.popup-header {
  display: flex;
}

.siblings {
  margin-top: 16px;
  list-style: none;
  padding: 0;
}

.sibling {
  display: block;
  white-space: normal;
  & + & {
    margin-top: 8px;
  }
  .active {
    color: colors.$gray24;
  }
}

.sibling-placeholder {
  background: colors.$gray4;
  border-radius: 4px;
  height: 12px;
  & + & {
    margin-top: 8px;
  }
}

.fade-enter {
  opacity: 0;
}

.fade-enter-active {
  opacity: 1;
  transition: opacity .2s ease-in-out;
}

.fade-exit {
  opacity: 1;
}

.fade-exit-active {
  opacity: 0;
  transition: opacity .2s ease-in-out;
}
