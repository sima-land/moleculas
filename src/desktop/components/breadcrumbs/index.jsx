import React, { useState, useRef, useEffect } from 'react';
import CSSTransition from 'react-transition-group/CSSTransition';
import Popup from '@dev-dep/ui-nucleons/popups/popup';
import on from '@dev-dep/ui-nucleons/helpers/on';
import classnames from 'classnames/bind';
import prop from 'lodash/fp/prop';
import DownSVG from './square-arrow-down.svg';
import styles from './breadcrumbs.scss';

const cx = classnames.bind(styles);

const DEFAULTS = {
  getName: prop('name'),
  getItemURL: prop('url'),
  getItemSiblings: prop('items'),
  isActiveItem: prop('isActive'),
  hasSiblings: breadcrumb => Array.isArray(breadcrumb?.items) && breadcrumb.items.length > 0,
};

/**
 * Список "хлебных крошек".
 * @param {Object} props Свойства компонента.
 * @param {Array} props.items Список "Хлебных крошек".
 * @param {Function} [props.getItemName] Вернёт имя элемента списка (по умолчанию свойство "name").
 * @param {Function} [props.getItemURL] Вернёт URL элемента списка (по умолчанию свойство "url").
 * @param {Function} [props.getItemSiblings] Вернёт список дочерних ссылок (по умолчанию свойство "items").
 * @param {Function} [props.isItemActive] Вернёт true если элемент представляет активный раздел.
 * @param {Function} [props.hasItemSiblings] Вернёт true если имеются соседние категории.
 * @param {Function} [props.onItemSiblingsShow] Будет вызвана при открытии списка соседних элементов.
 * @param {string} [props.'data-testid'] Идентификатор для систем автоматизированного тестирования.
 * @return {ReactElement} Элемент.
 */
export const Breadcrumbs = ({
  items,
  getItemName = DEFAULTS.getName,
  getItemURL = DEFAULTS.getItemURL,
  getItemSiblings = DEFAULTS.getItemSiblings,
  isItemActive = DEFAULTS.isActiveItem,
  hasItemSiblings = DEFAULTS.hasSiblings,
  onItemSiblingsShow,
  'data-testid': dataTestId,
}) => (
  <ul className={cx('breadcrumbs')} data-testid={dataTestId}>
    {Array.isArray(items) && items.map((breadcrumb, index) => (
      <Breadcrumb
        key={index}
        title={getItemName(breadcrumb)}
        href={getItemURL(breadcrumb)}
        canOpen={hasItemSiblings(breadcrumb)}
        siblings={getItemSiblings(breadcrumb)}
        active={isItemActive(breadcrumb)}
        getSiblingName={getItemName}
        isActiveSibling={isItemActive}
        getSiblingHref={getItemURL}
        onSiblingsShow={() => onItemSiblingsShow?.(breadcrumb)}
        data-testid={`breadcrumb:${index}`}
      />
    ))}
  </ul>
);

/**
 * Элемент списка "хлебных крошек".
 * @param {Object} props Свойства.
 * @return {ReactElement} Элемент.
 */
const Breadcrumb = ({
  title,
  href,
  siblings,
  active,
  getSiblingName,
  isActiveSibling,
  getSiblingHref,
  onSiblingsShow,
  canOpen,
  'data-testid': dataTestId,
}) => {
  const popupRef = useRef();
  const openerRef = useRef();
  const hasSiblings = siblings?.length > 0;
  const breadcrumbName = active
    ? (
      <span className={cx('active')}>{title}</span>
    ) : (
      <a
        className={cx('link')}
        href={href}
        children={title}
      />
    );

  const [withPopup, togglePopup] = useSiblingsPopup(popupRef, openerRef);

  return (
    <li data-testid={dataTestId}>
      {breadcrumbName}
      {canOpen && (
        <span ref={openerRef}>
          <DownSVG
            data-testid='breadcrumb:siblings-opener'
            className={cx('toggle-icon')}
            onClick={() => {
              togglePopup(true);
              onSiblingsShow?.();
            }}
          />
        </span>
      )}
      {hasSiblings && (
        <CSSTransition
          in={withPopup}
          timeout={200}
          unmountOnExit
          classNames={{
            enter: cx('fade-enter'),
            enterActive: cx('fade-enter-active'),
            exit: cx('fade-exit'),
            exitActive: cx('fade-exit-active'),
          }}
        >
          <Popup
            additionalClass={cx('popup')}
            ref={popupRef}
          >
            {breadcrumbName}
            <DownSVG
              data-testid='breadcrumb:siblings-closer'
              className={cx('toggle-icon', 'close')}
              onClick={() => togglePopup(false)}
            />
            <ul className={cx('siblings-list')}>
              {siblings.map((siblingItem, siblingIndex) => {
                const isActive = isActiveSibling(siblingItem);
                const siblingName = getSiblingName(siblingItem);

                return (
                  <li
                    key={siblingIndex}
                    className={cx('sibling-item')}
                    data-testid={`breadcrumbs-item-sibling:${siblingIndex}`}
                  >
                    {isActive ? (
                      <span className={cx('disabled')}>{siblingName}</span>
                    ) : (
                      <a
                        className={cx('link')}
                        color='dark-gray'
                        href={getSiblingHref(siblingItem)}
                        children={siblingName}
                      />
                    )}
                  </li>
                );
              })}
            </ul>
          </Popup>
        </CSSTransition>
      )}
    </li>
  );
};

/**
 * Хук работы с попапом соседних элементов "крошки".
 * @param {Object} popupRef Ref попапа.
 * @param {Object} openerRef Ref открывающего элемента.
 * @return {Function} Функция-обработчик монтирования компонента.
 */
const useSiblingsPopup = (popupRef, openerRef) => {
  const [withPopup, togglePopup] = useState(false);

  useEffect(() => {
    const offClick = on(window, 'click', ({ target }) => {
      target

        // если клик произошел не по узлу иконки открытия и не внутри popup'а
        && [popupRef.current, openerRef.current].every(node => node && node !== target && !node.contains(target))
        && togglePopup(false);
    });

    const offKeyDown = on(window, 'keydown', ({ code, keyCode }) => {
      (code === 'Escape' || keyCode === 27) && togglePopup(false);
    });

    return () => {
      offClick();
      offKeyDown();
    };
  }, []);

  return [withPopup, togglePopup];
};
