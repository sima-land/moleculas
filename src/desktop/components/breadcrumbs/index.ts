export { type BreadcrumbsProps, Breadcrumbs } from './breadcrumbs';
