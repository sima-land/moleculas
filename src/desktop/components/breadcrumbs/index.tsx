import React, { useEffect, useState, useRef } from 'react';
import CSSTransition from 'react-transition-group/CSSTransition';
import { Plate } from '@sima-land/ui-nucleons/plate';
import { Link } from '@sima-land/ui-nucleons/link';
import { times } from 'lodash';
import classnames from 'classnames/bind';
import styles from './breadcrumbs.module.scss';
import DownSVG from './square-arrow-down.svg';
import on from '@sima-land/ui-nucleons/helpers/on';

export interface Sibling {
  name: string;
  url?: string;
  [key: string]: any;
}

export interface Breadcrumb extends Sibling {
  siblings?: {
    state: 'initial' | 'fetching' | 'ready';
    data?: Sibling[];
  };
}

export interface BreadcrumbsProps {
  /** Список хлебных крошек. */
  items: Breadcrumb[];

  /** Получив хлебную крошку должна вернуть true если крошка выбрана. */
  isSiblingActive?: (sibling: Sibling) => boolean;

  /** Сработает при открытии попапа со смежными ссылками. */
  onSiblingsPopupOpen?: (item: Breadcrumb) => void;

  /** Идентификатор для систем автоматизированного тестирования. */
  'data-testid'?: string;
}

const cx = classnames.bind(styles);

/**
 * Список "хлебных крошек".
 * @param props Свойства.
 * @return Элемент.
 */
export const Breadcrumbs = ({
  items,
  onSiblingsPopupOpen,
  isSiblingActive,
  'data-testid': testId = 'breadcrumbs',
}: BreadcrumbsProps) => (
  <nav className={cx('root')} data-testid={testId}>
    {items.map((breadcrumb, index) => (
      <Breadcrumb
        key={index}
        data={breadcrumb}
        onSiblingsPopupOpen={() => onSiblingsPopupOpen?.(breadcrumb)}
        isSiblingActive={sibling => Boolean(isSiblingActive?.(sibling))}
      />
    ))}
  </nav>
);

/**
 * Элемент списка "хлебных крошек".
 * @param props Свойства.
 * @return Элемент.
 */
const Breadcrumb = ({
  data,
  onSiblingsPopupOpen,
  isSiblingActive,
}: {
  data: Breadcrumb;
  isSiblingActive: (sibling: Breadcrumb) => boolean;
  onSiblingsPopupOpen: () => void;
}) => {
  const popupRef = useRef<HTMLDivElement>(null);

  const [withPopup, togglePopup] = useState<boolean>(false);

  // с useOutsideClick не работает после обновления React до 17 версии, пока так
  useEffect(() => {
    if (withPopup) {
      return on(window, 'mousedown', togglePopup.bind(null, false));
    }
  }, [withPopup]);

  const selfName = (
    <Link color='gray87' href={data.url}>
      {data.name}
    </Link>
  );

  return (
    <div data-testid='breadcrumb' className={cx('breadcrumb')}>
      {selfName}

      {data.siblings && (
        <>
          <DownSVG
            data-testid='breadcrumb:siblings-opener'
            className={cx('toggle-icon')}
            onClick={() => {
              togglePopup(true);
              onSiblingsPopupOpen();
            }}
          />

          <CSSTransition
            in={withPopup}
            timeout={200}
            unmountOnExit
            classNames={{
              enter: cx('fade-enter'),
              enterActive: cx('fade-enter-active'),
              exit: cx('fade-exit'),
              exitActive: cx('fade-exit-active'),
            }}
          >
            <Plate ref={popupRef} shadow='z3' rounds='m' className={cx('popup')}>
              <div className={cx('popup-header')}>
                {selfName}
                <DownSVG
                  data-testid='breadcrumb:siblings-closer'
                  className={cx('toggle-icon', 'close')}
                  onClick={() => togglePopup(false)}
                />
              </div>

              <ul className={cx('siblings')}>
                {data.siblings.state === 'fetching' && (
                  <>
                    {times(8).map(index => (
                      <div key={index} className={cx('sibling-placeholder')} />
                    ))}
                  </>
                )}

                {data.siblings.state === 'ready' && (
                  <>
                    {data.siblings.data?.map((sibling, index) => (
                      <li key={index} className={cx('sibling')} data-testid='breadcrumb-sibling'>
                        {isSiblingActive(sibling) ? (
                          <span className={cx('active')}>{sibling.name}</span>
                        ) : (
                          <Link
                            color='gray87'
                            className={cx('link')}
                            href={sibling.url}
                            children={sibling.name}
                          />
                        )}
                      </li>
                    ))}
                  </>
                )}
              </ul>
            </Plate>
          </CSSTransition>
        </>
      )}
    </div>
  );
};
