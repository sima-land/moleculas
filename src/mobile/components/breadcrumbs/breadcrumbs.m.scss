@use 'pkg:@sima-land/ui-nucleons/colors.scss';

.root {
  display: flex;
  flex-wrap: wrap;
  margin-bottom: -4px;
  line-height: 16px;
  font-size: 12px;
}

.item {
  white-space: nowrap;
  display: block;
  margin: 0 4px 4px 0;
  overflow: hidden;
  text-overflow: ellipsis;
  &:not(:last-child)::after {
    content: '/';
    margin-left: 4px;
    color: colors.$basic-gray87;
  }
}
