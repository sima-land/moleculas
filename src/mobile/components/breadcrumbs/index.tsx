import { Link } from '@sima-land/ui-nucleons/link';
import React from 'react';
import styles from './breadcrumbs.module.scss';

export interface BreadcrumbsProps {
  items: {
    name: string
    url: string
  }[]
}

export const Breadcrumbs = ({ items }: BreadcrumbsProps) => (
  <nav className={styles.root}>
    {items.map((item, index) => (
      <Link
        key={index}
        href={item.url}
        color='gray87'
        className={styles.item}
      >
        {item.name}
      </Link>
    ))}
  </nav>
);
