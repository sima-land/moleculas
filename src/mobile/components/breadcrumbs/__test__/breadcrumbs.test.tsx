import { render } from '@testing-library/react';
import { Breadcrumbs } from '../breadcrumbs';

describe('Breadcrumbs', () => {
  it('should renders correctly', () => {
    const { container } = render(
      <Breadcrumbs
        items={[
          {
            name: 'Каталог',
            url: '#',
          },
          {
            name: 'Одежда и обувь',
            url: '#',
          },
          {
            name: 'Женская одежда',
            url: '#',
          },
          {
            name: 'Аксессуары',
            url: '#',
          },
          {
            name: 'Брендированные стойки «Этель» и «Доляна» в подарок за покупку кухонного текстиля',
            url: '#',
          },
          {
            name: 'Головные уборы',
            url: '#',
          },
          {
            name: 'Кепки, бейсболки',
            url: '#',
          },
          {
            name: 'Текущая страница',
            url: '#',
          },
        ]}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render item without url properly', () => {
    const { queryAllByTestId } = render(
      <Breadcrumbs
        items={[
          {
            name: 'First breadcrumb',
            url: 'https://first.breadcrumb/',
          },
          {
            name: 'Second breadcrumb',
            url: undefined,
          },
        ]}
      />,
    );

    const [first, second] = queryAllByTestId('breadcrumb');

    expect(first.getAttribute('href')).toBe('https://first.breadcrumb/');
    expect(first.getAttribute('role')).toBe(null);

    expect(second.getAttribute('href')).toBe(null);
    expect(second.getAttribute('role')).toBe('button');
  });
});
