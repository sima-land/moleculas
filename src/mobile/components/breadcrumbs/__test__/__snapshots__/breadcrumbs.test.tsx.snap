// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs should renders correctly 1`] = `
<div>
  <nav
    class="root"
  >
    <a
      class="link item basic-gray87"
      data-testid="breadcrumb"
      href="#"
    >
      Каталог
    </a>
    <a
      class="link item basic-gray87"
      data-testid="breadcrumb"
      href="#"
    >
      Одежда и обувь
    </a>
    <a
      class="link item basic-gray87"
      data-testid="breadcrumb"
      href="#"
    >
      Женская одежда
    </a>
    <a
      class="link item basic-gray87"
      data-testid="breadcrumb"
      href="#"
    >
      Аксессуары
    </a>
    <a
      class="link item basic-gray87"
      data-testid="breadcrumb"
      href="#"
    >
      Брендированные стойки «Этель» и «Доляна» в подарок за покупку кухонного текстиля
    </a>
    <a
      class="link item basic-gray87"
      data-testid="breadcrumb"
      href="#"
    >
      Головные уборы
    </a>
    <a
      class="link item basic-gray87"
      data-testid="breadcrumb"
      href="#"
    >
      Кепки, бейсболки
    </a>
    <a
      class="link item basic-gray87"
      data-testid="breadcrumb"
      href="#"
    >
      Текущая страница
    </a>
  </nav>
</div>
`;
